File: Simulation Code Matlab/Simulation Program Back APM/cruisecontroller.m
function  [Vd_APM, Vi_APM, switch_var, ti, j, alpha, v_ref, Vd_last_APM,t_interval, switch_intervall_bool] = cruisecontroller(t, t_start, expedite_t, expedite_p, d_segments, Vd_tab, p_APM, Vd_APM, Vi_APM, v_APM, Vd_last_APM, switch_var, ti, j, alpha, v_ref, k_a, k_d, s_a, s_d, t_interval,interval_bool, t_segments, switch_intervall_bool)
% Returns refrence velocity
%
% Inputs:
% t             = Time                                              (s)
% t_start       = Time APM useses to start up                       (s)
% expedite_t    = Time constant for expediting reference changes    (s)
% expedite_p    = Constant calculated from the expediting           (m)
% d_segments    = Array of input distance segments                  (s)
% Vd_tab        = Array of desired velocities for each segment      (m/s)
% p_APM         = Current position                                  (m)
% Vd_APM        = Desired velocity for the current segment          (m/s)
% Vi_APM        = Initial velocity before accelerating              (m/s)
% v_APM         = Current APM velocity                              (m/s)
% Vd_last_APM   = Last desired velocity                             (m/s)
% switch_var    = Switch parameter for diffent phases               (-)
% ti            = Initial time of the current segment               (s)
% j             = Index of the current segment in d_segments        (-)
% alpha         = Parameter for lowpassfilter                       (-)
% v_ref         = Reference velocity                                (m/s)
% k_a           = Sigmoid steepness parameter for acceleration      (1/s)
% k_d           = Sigmoid steepness parameter for deceleration      (1/s)
% s_a           = Sigmoid shift parameter for acceleration          (-)
% s_d           = Sigmoid shift parameter for deceleration          (-)
%
% Outputs:
% Vd_APM        = Updated desired velocity for the current segment  (m/s)
% Vi_APM        = Updated initial velocity                          (m/s)
% switch_var    = Updated switch parameter                          (-)
% ti            = Updated start time of the current segment         (s)
% j             = Updated index of the current segment in d_segments(-)
% alpha         = Updated acceleration or deceleration parameter    (-)
% v_ref         = Reference velocity                                (m/s)
% Vd_last_APM   = Last desired velocity                             (m/s)
%
% Author:    Camilla Kvamme
% Date:      13.03.2024
%

% Start CC after the startup time 

    if (t > t_start - expedite_t &&  j ~= length(Vd_tab) + 1) 
            % If the position is larger than the set sqedule 
            % then update wanted speed (Vd_APM) for the APM.
            if ~interval_bool && (p_APM >= d_segments(j) - expedite_p(j)) 
                Vd_last_APM = Vd_APM;
                if(j < length(Vd_tab))
                    Vd_APM = Vd_tab(j+1);
                end
                Vi_APM = v_APM; % Set new initial velocity
                
                % Based on the new wanted speed (Vd_APM) 
                % decide witch phase the APM is in.
                if (Vd_APM > Vd_last_APM)
                    switch_var = 1;         % Acceleration phase
                    ti = t;                 % Update time initial time
                elseif (Vd_APM == 0)
                    switch_var = 4;         % Constant position phase
                elseif (Vd_APM < Vd_last_APM)
                    switch_var = 3;         % Deceleration phase
                    ti = t;                 % Update time initial time     
                elseif (Vd_APM <= Vi_APM + 0.001) && (Vd_APM >= Vi_APM - 0.001)
                    switch_var = 2;         % Constant velocity phase
                end
                % Update j 
                j = j + 1; 
                % Updating alpha for lowpass 
                alpha = 1;

            elseif interval_bool && switch_intervall_bool
                Vd_last_APM = Vd_APM;
                if(j < length(Vd_tab))
                    Vd_APM = Vd_tab(j+1);
                end
                Vi_APM = v_APM; % Set new initial velocity
                
                % Based on the new wanted speed (Vd_APM) 
                % decide witch phase the APM is in.
                if (Vd_APM > Vd_last_APM)
                    switch_var = 1;         % Acceleration phase
                    ti = t;                 % Update time initial time
                elseif (Vd_APM == 0)
                    switch_var = 4;         % Constant position phase
                elseif (Vd_APM < Vd_last_APM)
                    switch_var = 3;         % Deceleration phase
                    ti = t;                 % Update time initial time     
                elseif (Vd_APM <= Vi_APM + 0.001) && (Vd_APM >= Vi_APM - 0.001)
                    switch_var = 2;         % Constant velocity phase
                end

                % Update counter for the list of Vd_APM and time 
                j = j + 1; 
                switch_intervall_bool = false;

                % Updating alpha for lowpass 
                alpha = 1;
            end           
    end
    % Different phases in CC
    switch switch_var
        case 1 % First phase: acceleration
            [v_ref] = acceleration_phase( t, ti, Vi_APM, Vd_APM, k_a, s_a);
        case 2
           % Second phase: constant velocity
            v_ref = Vd_APM;
        case 3 % Third phase: deceleration  
            [v_ref] = acceleration_phase( t, ti, Vi_APM, Vd_APM, k_d, s_d); 

        case 4 % Forth phase: constant position    
            v_ref = 0;
    
        otherwise
            % Handle any other cases if needed
            display('Error');
    end
end

File: Python/speed_reference/cruise_controller.py
import time
import numpy as np

# Intialize parameters
j = 0           # Counter for updating CC parameters
ti_CC = 0       # Initial time for start of CC phase
Vi_CC = 0       # Initial velocity for start of CC pahse 
Vd_APM = 0      # Current desirec velocity
Vd_last = 0     # Last value of Vd

# Sigmoid parameters
k_a = 1.5       # Sigmoid steepness parameter for acceleration (m/s)
k_d = 1.5       # Sigmoid steepness parameter for deceleration (m/s)
s_a = 7         # Sigmoid shift parameter for acceleration
s_d = 7         # Sigmoid shift parameter for deceleration
expedite_t = 7  # Expedite time of the acceleration

def acceleration_phase(t_i, v_i, v_d, k, s):

    '''
    t_i: initial time
    v_i: inital velocity
    v_d: desired velocity
    k:   sigmoid steepness parameter
    s:   sigmoid shift parameter
    '''
    t = time.time()

    sigmoid = 1 / (1 + np.exp(-k * ((t - t_i) - s)))

    v_sigmoid = v_i + (v_d - v_i)*sigmoid

    return v_sigmoid

def updateCCparameters(Vd_list, p_segment_list, total_pos_travelled, speed_APM):
    
    global j, Vd_APM, Vi_CC, ti_CC, Vd_last 

    if j < len(Vd_list) and (total_pos_travelled >= p_segment_list[j] - Vd_list[j]*expedite_t):
        j+= 1
        Vd_last = Vd_APM
        Vd_APM = Vd_list[j]
        Vi_CC = speed_APM
        ti_CC = time.time()


def cruise_controller():
    
    if Vd_APM > Vi_CC:
        phase_var = 'acceleration'
    elif Vd_APM < Vi_CC:
        phase_var = 'deceleration'
    elif Vd_APM == 0:
        phase_var = 'constant_position'
    else:
        phase_var = 'constant_velocity'
    
    match phase_var:
        case 'acceleration': # First phase: acceleration
            cc_speed_ref = acceleration_phase(ti_CC, Vi_CC, Vd_APM, k_a, s_a)
        case 'constant_velocity': # Second phase: constant velocity
            cc_speed_ref = Vd_APM
        case 'deceleration': # Third phase: deceleration
            cc_speed_ref = acceleration_phase(ti_CC, Vi_CC, Vd_APM, k_d, s_d)
        case 'constant_position': # Fourth phase: constant positiion
            cc_speed_ref = 0
        case _:
            print('Error')
    if(cc_speed_ref < 0):
        cc_speed_ref = 0

    return cc_speed_ref

    
